### PR TITLE
Datepicker locale fix

### PR DESCRIPTION
**Pull request by Blackbaud-AlexKingman**
Addresses #139 

When trying to set or override the datepicker's date formatting, it should follow the rules of specificity. For example, datepicker will now try to set dateformat in the following order:
1) the `@Input()` property of `dateformat`
2) the `dateFormat` property on the `SkyDatepickerConfigService`
3) the preferred short date format from the locale established in the `SkyAppLocaleProvider`
4) if nothing of these are found, default to `MM/DD/YYYY`

### CODE REVIEW

File: src/app/public/modules/datepicker/datepicker-calendar.component.ts [hunk 1]
```diff
@@ -70,17 +70,15 @@ export class SkyDatepickerCalendarComponent implements AfterViewInit {
...
-  protected config: SkyDatepickerConfigService;
 
   private formatter = new SkyDateFormatter();
 
   private _startingDay: number;
 
   public constructor(
     private adapter: SkyDatepickerAdapterService,
-    config: SkyDatepickerConfigService,
+    private config: SkyDatepickerConfigService,
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/datepicker/datepicker-calendar.component.ts:80]: ☝️ This was just for consistency's sake.

File: src/app/public/modules/timepicker/timepicker.component.spec.ts [hunk 2]
```diff
@@ -547,8 +547,7 @@ describe('Timepicker', () => {
       expect(getInput(fixture)).not.toHaveCssClass('ng-invalid');
     }));
 
-    // TODO: This should be fixed with issue https://github.com/blackbaud/skyux-datetime/issues/135
-    xit('should properly update model and input when required and undefined', fakeAsync(() => {
+    it('should properly update model and input when required and undefined', fakeAsync(() => {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/timepicker/timepicker.component.spec.ts:550]: Unrelated change to this PR. Just figured out this was working as expected - just needed some more `ticks` in there.

File: src/app/visual/datepicker/datepicker-visual.component.html [hunk 3]
```diff
@@ -83,7 +83,6 @@
         [minDate]="minDate"
         [skyDatepickerNoValidate]="noValidate"
         [startingDay]="startingDay"
-        [disabled]="disabled"
```
Comment (Blackbaud-AlexKingman) [inline src/app/visual/datepicker/datepicker-visual.component.html:86]: Cleaned up this visual test so we would no longer get the console warning messages.

File: src/app/public/modules/datepicker/fuzzy-datepicker.component.spec.ts [hunk 4]
```diff
@@ -47,6 +44,8 @@ import {
 import {
   FuzzyDatepickerReactiveTestComponent
 } from './fixtures/fuzzy-datepicker-reactive.component.fixture';
+import { SkyAppLocaleProvider } from '@skyux/i18n';
+import { Observable } from 'rxjs';
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/datepicker/fuzzy-datepicker.component.spec.ts:48]: Code style, import statements.

File: src/app/public/modules/datepicker/fuzzy-datepicker.component.spec.ts [hunk 5]
```diff
@@ -4,7 +4,8 @@ import {
   fakeAsync,
   flush,
   TestBed,
-  tick
+  tick,
+  inject
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/datepicker/fuzzy-datepicker.component.spec.ts:8]: Code style, alphabetical ordering.

File: src/app/public/modules/datepicker/datepicker.component.spec.ts [hunk 6]
```diff
@@ -1224,47 +1250,29 @@ describe('datepicker', () => {
...
     it('should display formatted date based on locale by default', fakeAsync(() => {
-      spyOn(mockWindowService, 'getWindow').and.returnValue(
-        {
-          navigator: {
-            languages: ['es'] // Spanish
-          }
-        }
+      spyOn(localProvider, 'getLocaleInfo').and.returnValue(
+        Observable.of({
+              locale: 'es' // Set locale to Spanish.
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/datepicker/datepicker.component.spec.ts:1265]: Code style, tabs.

File: src/app/public/modules/datepicker/datepicker.component.spec.ts [hunk 7]
```diff
@@ -3,7 +3,8 @@ import {
   ComponentFixture,
   fakeAsync,
   TestBed,
-  tick
+  tick,
+  inject
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/datepicker/datepicker.component.spec.ts:7]: Code style, alphabetical ordering.

File: src/app/public/modules/datepicker/datepicker-input-fuzzy.directive.ts [hunk 8]
```diff
@@ -251,13 +265,21 @@ export class SkyFuzzyDatepickerInputDirective
...
+    private localeProvider: SkyAppLocaleProvider,
+    private renderer: Renderer2,
+    private resourcesService: SkyLibResourcesService,
+    @Optional() private datepickerComponent: SkyDatepickerComponent
+  ) {
+    this.localeProvider.getLocaleInfo()
+      .takeUntil(this.ngUnsubscribe)
+      .subscribe((localeInfo) => {
+        moment.locale(localeInfo.locale || 'en-US');
+        this.localeProviderDateFormat = moment.localeData().longDateFormat('L');
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/datepicker/datepicker-input-fuzzy.directive.ts:280]: It looks like this same locale logic is being used in three different places. Is there a universal location we could put this code so that if we fix it in one place, all three directives benefit?

File: src/app/public/modules/datepicker/datepicker.component.spec.ts (hunk 7 above)
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/datepicker.component.spec.ts:7]: Fixed.

File: src/app/public/modules/datepicker/datepicker.component.spec.ts (hunk 6 above)
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/datepicker.component.spec.ts:1265]: Fixed.

File: src/app/public/modules/datepicker/fuzzy-datepicker.component.spec.ts (hunk 5 above)
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/fuzzy-datepicker.component.spec.ts:8]: Fixed.

File: src/app/public/modules/datepicker/fuzzy-datepicker.component.spec.ts (hunk 4 above)
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/fuzzy-datepicker.component.spec.ts:48]: Fixed.

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-datetime?_sp.spa=pr-datepicker-locale-fix](https://developer.blackbaud.com/skyux-datetime?_sp.spa=pr-datepicker-locale-fix)

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-datetime?_sp.spa=pr-datepicker-locale-fix](https://developer.blackbaud.com/skyux-datetime?_sp.spa=pr-datepicker-locale-fix)

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-datetime?_sp.spa=pr-datepicker-locale-fix](https://developer.blackbaud.com/skyux-datetime?_sp.spa=pr-datepicker-locale-fix)

**Comment by Blackbaud-BobbyEarl** [comment]
Preview build at [https://developer.blackbaud.com/skyux-datetime?_sp.spa=pr-datepicker-locale-fix](https://developer.blackbaud.com/skyux-datetime?_sp.spa=pr-datepicker-locale-fix)
